JIT: Add 3-opt implementation for improving upon RPO-based layout

amanasifkhalid opened the pull request:
Part of #107749. After establishing a baseline layout, where hot blocks are in relative reverse post-order and cold blocks are out-of-line, run a [3-opt](https://en.wikipedia.org/wiki/3-opt) pass over each try region, as well as the main method body. Profitability of each partition is currently measured with edge weights only, without considering the type ([un]conditional, forward/backward, etc) of the branch. To minimize the search space of partition points, we maintain a priority queue of flow edges between non-contiguous blocks, such that we can attempt to create fallthrough for the next most-profitable edge.

To avoid disturbing EH semantics too much, we currently only reorder blocks within EH regions. This could be relaxed somewhat to at least move region entries up to their hottest predecessors, effectively moving the entire region once we re-establish contiguity -- I plan to try this in a follow-up PR. Note that we don't consider reordering handler regions beyond the initial RPO-based layout, under the assumption that the overhead of exception handling usually minimizes the potential gains from better layout.

amanasifkhalid committed:
Implement k-opt for non-EH methods

amanasifkhalid committed:
Enable for methods with EH

amanasifkhalid committed:
Merge branch 'main' into k-opt-layout

amanasifkhalid committed:
Add comments

amanasifkhalid committed:
Style

amanasifkhalid committed:
Merge branch 'main' into k-opt-layout

amanasifkhalid committed:
Only one iteration for now; try to reduce TP cost

amanasifkhalid committed:
Remove initial layout cost calculation

amanasifkhalid committed:
Conditionalize EH checks

AndyAyersMS commented on src/coreclr/jit/fgopt.cpp:
```diff
@@ -5106,6 +5123,236 @@ void Compiler::fgMoveColdBlocks()
+        improvedLayout              = false;
+        BasicBlock* const exitBlock = blockVector[blockCount - 1];
+
+        for (unsigned i = 1; i < (blockCount - 1); i++)
```
I think the root of the TP cost is here -- we want to avoid having to search for possible cut points.

One approach is to just pick randomly, but I think we can do better for now. Roughly speaking in the pass above we should find all blocks that are either not just before their optimal successor and/or not just after their optimal successor.

We can rank these by the difference in the current vs optimal score. Then greedily pick the worst, that gives the first cut point. For the second cut point you can pick the best pred for the first cut point's current next block, or the best succ for the current pred of the first cut point's ideal successor. That is, if we have

```
S ~~~~ 1|2 ~~~ 3|4 ~~~ 5|6 ~~~ E

1's ideal succ is 4

reordering is

S ~~~~ 1|4 ~~~ 5|2 ~~~ 3|6 ~~~ E

So we either try and find a 5 which is the ideal pred of 2, or a 6 which is the ideal succ of 3.

Failing that we might pick some other block that is not currently followed by its ideal succ.
```
So one idea is to keep 3 values for each block: its min score, current score, and best score (lower is better). Order the blocks by current-min. Pick of the best as the first split, and then see if any of the next few provide a good second split.

Likely though this ends up needing a priority queue or similar as once we accept an arrangement we need to update some of the costings...

amanasifkhalid committed:
Merge from main

amanasifkhalid committed:
Add priority queue impl

amanasifkhalid committed:
wip

amanasifkhalid committed:
Fix lambda capture

amanasifkhalid committed:
Merge branch 'main' into k-opt-layout

amanasifkhalid committed:
Consider forward conditional jumps

amanasifkhalid committed:
Remove debug print

amanasifkhalid committed:
Consider backward jumps; find more initial candidates

amanasifkhalid committed:
Revert irrelevant changes

amanasifkhalid committed:
Missed a few

amanasifkhalid committed:
Add JitDump check

amanasifkhalid committed:
Add more candidate edges when reordering

amanasifkhalid committed:
Merge branch 'main' into k-opt-layout

amanasifkhalid committed:
Don't add duplicate edges to cutPoints

amanasifkhalid committed:
Consider each candidate edge at most once

amanasifkhalid commented:
@AndyAyersMS thanks for bearing with me on this. I've implemented your suggestion of building and maintaining a priority queue of cut points, and this seems to be sufficiently cheap. [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=820809&view=ms.vss-build-web.run-extensions-tab) show plenty of variance in asmdiffs across platforms, though this looks like a net PerfScore win. To contain the number of iterations, we currently consider each edge at most once; we probably don't want to limit the search space too much, though these limitations had pretty small diffs locally, so it seems like the current approach is fixing the most obvious instances of subpar layout.

I haven't implemented this for methods with EH yet, though I'm thinking of leaving the cutpoint search as-is, and then after reordering blocks, we can make EH regions contiguous by "bubbling up" the next EH block we see to its predecessor. This fixup can break up fallthrough from EH exits into non-EH blocks, but it will maintain the relative ordering such that the exit jump is forward; for now, breaking up such fallthrough seems necessary. With this approach, we can get rid of the EH fixup logic in earlier ordering passes (RPO layout, `fgMoveColdBlocks`, etc) and win back some TP -- but I wanted to evaluate that separately from this PR.

I think this PR is in good shape, so I thought I'd ping you now in case you want to take a look, though I don't plan to push to merge this until we get the LSRA changes where we want them.

amanasifkhalid committed:
Merge from main

amanasifkhalid committed:
Don't factor cold branches into cost calculatioN

amanasifkhalid committed:
Merge from main

amanasifkhalid committed:
Remove used candidates set

amanasifkhalid committed:
Revert "Remove used candidates set"

This reverts commit d6fea98ce7a0a73e13f70f9de97da55bbbd99d7d.

amanasifkhalid committed:
Tweak cold block finding

amanasifkhalid committed:
Refactor into ThreeOptLayout class

amanasifkhalid committed:
Reorder EH regions

amanasifkhalid committed:
Fix completely cold try regions (clean replay)

amanasifkhalid committed:
Small cleanup

amanasifkhalid committed:
Add currEHRegion member

amanasifkhalid committed:
Only move blocks within same region after 3-opt

amanasifkhalid committed:
Merge from main

amanasifkhalid committed:
Cleanup

amanasifkhalid committed:
Comments

amanasifkhalid commented:
/azp run runtime-coreclr outerloop, Fuzzlyn, Antigen

amanasifkhalid commented:
cc @dotnet/jit-contrib, @AndyAyersMS PTAL. [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=855874&view=ms.vss-build-web.run-extensions-tab) are a bit broken on Linux at the moment, though other platforms show less dramatic churn than I expected -- it's possible the current cost model isn't sophisticated enough to incentivize larger bets. PerfScore diffs look like a net improvement, and TP cost is pretty cheap. Antigen and Fuzzlyn failures don't look related to layout. Thanks!

AndyAyersMS reviewed (commented):
Some preliminary notes.

I find it a little confusing to use "cost" if the goal of the optimization is to maximize. Maybe consider "benefit"? Or rework things to be minimizing cost?

I also think it would be useful to build the overall score for a layout; you should be able to show at each step this value is increasing (/decreasing if you minimize).

When we reorder it seems like there are six impacted blocks (3 have new preds, 3 have new succ), so I'm surprised we are just reconsidering 3 sets of edges. Maybe this is related to not wanting to mess with the boundaries? But even so I'd expect to see 4 sets...

I would recommend adding a doc with some simple examples worked out in pseudocode.

amanasifkhalid committed:
EdgeCmp tie-breaker

amanasifkhalid commented on src/coreclr/jit/fgopt.cpp:
```diff
@@ -4855,6 +4855,467 @@ void Compiler::fgMoveColdBlocks()
+{
+    assert(edge != nullptr);
+
+    BasicBlock* const srcBlk = edge->getSourceBlock();
```
> Would it make sense to keep track of edges that can never be handled? Perhaps a second tracked set?

I suppose I could just move the set insertion/membership check to the beginning of `ConsiderEdge`? This is a no-diff change for me locally, and it seems like a convenient way to short-circuit all the checks.

> Alternatively, we could keep some state on the edge itself, there is likely some padding void there already that could hold bits.

That sounds worth pursuing. I'll try this out.

amanasifkhalid committed:
Comment feedback; replace 'usedCandidates' set with FlowEdge flag

amanasifkhalid committed:
Reframe cost model as maximal score problem

amanasifkhalid committed:
Small refactor

amanasifkhalid commented:
@AndyAyersMS thanks for the review!

> I find it a little confusing to use "cost" if the goal of the optimization is to maximize. Maybe consider "benefit"? Or rework things to be minimizing cost?

Agreed that the semantics looked weird. I've reframed the partition "cost" as its "score," such that we want to maximize the score of a layout by creating fallthrough for the heaviest edges. Personally, I find the cost-minimizing framing of this problem more intuitive, but if/when the model deviates from just using edge weights, I think it will be easier to compute the benefit rather than the cost of an edge (assuming the cost is computed by subtracting the edge's score from all other flow out of the source block, which may not be the same as the source block's weight).

> I also think it would be useful to build the overall score for a layout; you should be able to show at each step this value is increasing (/decreasing if you minimize).

Do you intend for this to be a debug-only utility, i.e. to improve the JitDump info reported by 3-opt? Or is there some strategic advantage to knowing the total layout score during 3-opt?

> When we reorder it seems like there are six impacted blocks (3 have new preds, 3 have new succ), so I'm surprised we are just reconsidering 3 sets of edges. Maybe this is related to not wanting to mess with the boundaries? But even so I'd expect to see 4 sets...

Yeah, this bit of logic hasn't aged well over the last few iterations. I've simplified it a bit, and diffs against the last revision seem to show we're capturing more reordering opportunities.

> I would recommend adding a doc with some simple examples worked out in pseudocode.

Is it alright if I split that out into a separate PR?

amanasifkhalid committed:
Simplify adding new candidate edges

amanasifkhalid commented:
/azp run runtime-coreclr outerloop, Fuzzlyn, Antigen

AndyAyersMS commented:
> > I also think it would be useful to build the overall score for a layout; you should be able to show at each step this value is increasing (/decreasing if you minimize).
> 
> Do you intend for this to be a debug-only utility, i.e. to improve the JitDump info reported by 3-opt? Or is there some strategic advantage to knowing the total layout score during 3-opt?

Mostly debug, I guess. When we did this way back when we built the global scorer first and tried to ensure that a good score looked like a good layout, and that a bad score looked like a bad layout, and that there were sufficient methods with bad layouts (and a set of "bad" candidates to look at).

But if you know how close the score is to a good score, you can also use this for early stopping.

AndyAyersMS commented:
> > I would recommend adding a doc with some simple examples worked out in pseudocode.
> 
> Is it alright if I split that out into a separate PR?

Yes, sure.

AndyAyersMS reviewed (approved):
Looks good!

amanasifkhalid commented on src/coreclr/jit/fgopt.cpp:
```diff
@@ -4855,6 +4855,483 @@ void Compiler::fgMoveColdBlocks()
+    unsigned position = 0;
+
+    // Initialize current block order
+    for (BasicBlock* const block : compiler->Blocks(compiler->fgFirstBB, finalBlock))
```
Splitting the loops up allows us to calculate the exact size of `blockOrder`/`tempOrder` before they're needed. Instead of allocating the exact size needed, I can use `fgBBcount` for the array sizes (and maybe trim this upper bound as we search for the last hot block). It'll waste some memory, but then we only need one loop.

amanasifkhalid committed:
Guess number of hot blocks when allocating arrays

amanasifkhalid committed:
Skip reordering if too few blocks

amanasifkhalid committed:
Add another check for too few blocks

amanasifkhalid committed:
JITDUMP msg

amanasifkhalid commented:
@AndyAyersMS thanks for the reviews! Are you ok with this revision?

AndyAyersMS reviewed (approved):
Yes

LoopedBard3 commented:
Issue tracking regressions from this PR (since GitHub doesn't seem to want to do autolinks at the moment): https://github.com/dotnet/runtime/issues/109613
